hat/backends/spirv/src/main/java/intel/code/spirv/SpirvModuleGenerator.java

```diff
@@ -180,6 +180,214 @@ private MemorySegment generateModuleInternal(SpirvOp.FuncOp func) {
         return finalizeModule();
     }
 
+    private int getTypeSize(String typeName) {
+        if (sizeMap.containsKey(typeName)) {
+            return sizeMap.get(typeName);
+        }
+        switch (typeName) {
+            case "byte" -> {
+                return 1;
+            }
+            case "boolean" -> {
+                return 1;
+            }
+            case "short" -> {
+                return 2;
+            }
+            case "int" -> {
+                return 4;
+            }
+            case "long" -> {
+                return 8;
+            }
+            case "float" -> {
+                return 4;
+            }
+            case "double" -> {
+                return 8;
+            }
+            default -> {
+                throw new IllegalStateException("unknown type " + typeName);
+            }
+        }
+    }
+
+    private void addTypeToModule(String name, SPIRVId typeIdsArray[]) {
+        String upperName = name.substring(0, 1).toUpperCase() + name.substring(1);
+        String ptrUpperName = "ptr" + upperName;
+        String ptrPtrUpperName = "ptrPtr" + upperName;
+        module.add(new SPIRVOpTypeStruct(nextId(name), new SPIRVMultipleOperands<>(typeIdsArray)));
+        moduleTypes.add(name);
+        module.add(new SPIRVOpTypePointer(nextId(ptrUpperName), SPIRVStorageClass.CrossWorkgroup(), getType(name)));
+        moduleTypes.add(ptrUpperName);
+        module.add(new SPIRVOpTypePointer(nextId(ptrPtrUpperName), SPIRVStorageClass.CrossWorkgroup(), getType(ptrUpperName)));
+        moduleTypes.add(ptrPtrUpperName);
+    }
+
+    private void addArrayToModule(String name, String typeName, int len) {
+        SPIRVOpConstant constant = new SPIRVOpConstant(getType("int"), nextId(), new SPIRVContextDependentInt(new BigInteger(String.valueOf(len))));
+        module.add(constant);
+        SPIRVOpTypeArray typeArray = new SPIRVOpTypeArray(nextId(name + "Array"), getType(typeName), constant.getResultId());
+        module.add(typeArray);
+        moduleTypes.add(name + "Array");
+        SPIRVOpTypePointer ptrTypeArray = new SPIRVOpTypePointer(nextId("ptr" + name + "Array"), SPIRVStorageClass.CrossWorkgroup(), getType(name + "Array"));
+        module.add(ptrTypeArray);
+        moduleTypes.add("ptr" + name + "Array");
+    }
+
+    private void generateTypeDeclaration(Object... args) {
```

Comment:
@grfrost can better advise, but if possible it might be better to work directly off the `MemoryLayout` objects used by the buffer, that's also a more general solution.